### PR TITLE
Added transformer for onUsingTick in EntityPlayer

### PR DESCRIPTION
**Pull request by RecursivePineapple**
EntityPlayer does not call onUsingTick properly on the server because it uses a == instead of a proper ItemStack.areItemStacksEqual call. Another piece of code constantly copies the itemInUse stack, so this condition is never true. This transformer changes the == to call ItemStack.areItemStacksEqual instead. This isn't a mixin because mixins can't change expressions like this.

### CODE REVIEW

**Comment by RecursivePineapple** [comment]
I've been testing this for a few weeks and it works well. I enabled the transformer on both sides (client & server) even though it's not necessary because the client only works by happenstance.

**Review by glowredman — CHANGES_REQUESTED** [review]
1. Please use the correct types instead of `var`
2. Please use a LOG4J logger instead of `System.out`

**Comment by Alexdoru** [comment]
> Another piece of code constantly copies the itemInUse stack

what other piece of code ?

File: src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        if (NameContext.Obf.entityPlayer.equals(name)) {
+            final ClassReader cr = new ClassReader(basicClass);
+            final ClassNode cn = new ClassNode();
+            cr.accept(cn, 0);
+            transformClassNode(NameContext.Obf, cn);
+            ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_MAXS);
+            cn.accept(cw);
+
+            LOGGER.info("Transformed {}", name);
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java:31]: you don't need to log that you are transforming a class, however you should use a try catch and if it returns an exception, you should log it and return the untransformed class

File: src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            return cw.toByteArray();
+        }
+
+        return basicClass;
+    }
+
+    public void transformClassNode(NameContext names, ClassNode cn) {
+        LOGGER.info("Found EntityPlayer (class = {}, name context = {})", names.entityPlayer, names.name());
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java:51]: don't need to log

File: src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+        for (var method : cn.methods) {
+            if (method.name.equals(names.onUpdate)) {
+                LOGGER.info("Found EntityPlayer.onUpdate ({})", names.onUpdate);
+
+                var cursor = method.instructions.getFirst();
+
+                while (cursor != null) {
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java:59]: why don't you use the built in iterator or a for loop over instructions.toArray()

File: src/main/java/com/mitchej123/hodgepodge/util/ASMUtil.java [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+import org.objectweb.asm.tree.InvokeDynamicInsnNode;
+import org.objectweb.asm.tree.LdcInsnNode;
+import org.objectweb.asm.tree.LineNumberNode;
+import org.objectweb.asm.tree.LocalVariableNode;
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MultiANewArrayInsnNode;
+import org.objectweb.asm.tree.TypeInsnNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+public class ASMUtil {
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/util/ASMUtil.java:23]: is this class even used ? just delete it

File: src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+            LOGGER.info("Transformed {}", name);
+            return cw.toByteArray();
+        }
+
+        if (NameContext.Deobf.entityPlayer.equals(name)) {
+            final ClassReader cr = new ClassReader(basicClass);
+            final ClassNode cn = new ClassNode();
+            cr.accept(cn, 0);
+            transformClassNode(NameContext.Deobf, cn);
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java:39]: I wouldn't do that to check the obfuscation state, check it from the injectData method in the `IFMLLoadingPlugin` implementation

**Review by Alexdoru — CHANGES_REQUESTED** [review]

File: src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+public class EntityPlayerTransformer implements IClassTransformer {
+
+    private static final Logger LOGGER = LogManager.getLogger("EntityPlayerTransformer");
+
+    @Override
+    public byte[] transform(String name, String transformedName, byte[] basicClass) {
+        if (basicClass == null) return null;
+
+        if (NameContext.Obf.entityPlayer.equals(name)) {
```
[the remark this answers is not in the document]
Comment (RecursivePineapple) [reply src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java:23]: Is only transformedName deobfuscated, or everything? If it's only transformedName, how should I determine what names to use?

File: src/main/java/com/mitchej123/hodgepodge/util/ASMUtil.java (hunk 4 above)
Comment (RecursivePineapple) [reply src/main/java/com/mitchej123/hodgepodge/util/ASMUtil.java:23]: It was useful for me, so I figured I should leave it in for others. Instructions don't have a toString implementation, so you can't see what instructions in an InsnList do without debugging the transformer. Should I still remove it?

Comment (Alexdoru) [reply src/main/java/com/mitchej123/hodgepodge/util/ASMUtil.java:23]: yeah remove it

File: src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (RecursivePineapple) [reply src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java:23]: I tested it in the full pack and it looks like this isn't the case. I had to use the obfuscated name.

[the remark this answers is not in the document]
Comment (Alexdoru) [reply src/main/java/com/mitchej123/hodgepodge/asm/transformers/mc/EntityPlayerTransformer.java:23]: you can do `"net.minecraft.entity.player.EntityPlayer".equals(transformedName)` even when in obfuscated environement

**Comment by Alexdoru** [comment]
this introduces a bug where if you have 2 of the same items in your hotbar for example 2 bows, you can draw your first bow and then switch the the second one and it will be fully draw, same with eating the same food

We should instead find what code changes the items while they are being used which prevents the vanilla code `==` from being true

**Pull request closed without merging**